src/backend/trace.rs

```diff
@@ -468,3 +468,229 @@ impl TraceData {
+    // #[test_case(file!("ipv4_3probes_3hops_completed.yaml"))]
+    // #[test_case(file!("ipv4_4probes_all_status.yaml"))]
+    // #[test_case(file!("ipv4_4probes_0latency.yaml"))]
+    fn test_scenario(scenario: Scenario) {
+        let mut trace = Trace::new(100, 1);
+        for (i, round) in scenario.rounds.into_iter().enumerate() {
+            let probes = round
+                .probes
+                .into_iter()
+                .map(|p| ProbeRound(p, Round(i)))
+                .map(Into::into)
+                .collect::<Vec<_>>();
+            let largest_ttl = TimeToLive(scenario.largest_ttl);
+            let tracer_round =
+                TracerRound::new(&probes, largest_ttl, CompletionReason::TargetFound);
+            trace.update_from_round(&tracer_round);
+        }
+        let actual_hops = trace.hops(Trace::default_flow_id());
+        let expected_hops = scenario.expected.hops;
+        for (actual, expected) in actual_hops.iter().zip(expected_hops) {
+            assert_eq!(actual.ttl(), expected.ttl);
+            assert_eq!(
+                actual.addrs().collect::<HashSet<_>>(),
+                expected.addrs.keys().collect::<HashSet<_>>()
+            );
+            assert_eq!(actual.addr_count(), expected.addrs.len());
+            assert_eq!(actual.total_sent(), expected.total_sent);
+            assert_eq!(actual.total_recv(), expected.total_recv);
+            assert_eq_optional(Some(actual.loss_pct()), Some(expected.loss_pct));
+            assert_eq_optional(actual.last_ms(), expected.last_ms);
+            assert_eq_optional(actual.best_ms(), expected.best_ms);
+            assert_eq_optional(actual.worst_ms(), expected.worst_ms);
+            assert_eq_optional(Some(actual.avg_ms()), Some(expected.avg_ms));
+            assert_eq_optional(actual.jitter_ms(), expected.jitter);
+            assert_eq_optional(Some(actual.javg_ms()), Some(expected.javg));
+            assert_eq_optional(actual.jmax_ms(), expected.jmax);
+            assert_eq_optional(Some(actual.jinta()), Some(expected.jinta));
+            assert_eq!(actual.last_src_port(), expected.last_src);
+            assert_eq!(actual.last_dest_port(), expected.last_dest);
+            assert_eq!(actual.last_sequence(), expected.last_sequence);
+            assert_eq!(
+                Some(
+                    actual
+                        .samples()
+                        .iter()
+                        .map(|s| s.as_secs_f64() * 1000_f64)
+                        .collect()
+                ),
+                expected.samples
+            );
+        }
+    }
+
+    #[allow(clippy::float_cmp)]
+    fn assert_eq_optional(actual: Option<f64>, expected: Option<f64>) {
+        match (actual, expected) {
+            (Some(actual), Some(expected)) => assert_eq!(actual, expected),
+            (Some(_), None) => panic!("actual {actual:?} but not expected"),
+            (None, Some(_)) => panic!("expected {expected:?} but no actual"),
```

Comment:
@trkelly23 tweaked so that we fail for cases where either expected or actual is missing.  We could relax this to allow cases where we have the actual without the expected to give us a way to "ignore" the field but I tend to prefer it being strict.